openmc/data/resonance.py

```diff
@@ -0,0 +1,1063 @@
+    energy_min : float
+        Minimum energy of the resolved resonance range in eV
+    energy_max : float
+        Maximum energy of the resolved resonance range in eV
+    channel : dict
+        Dictionary whose keys are l-values and values are channel radii as a
+        function of energy
+    scattering : dict
+        Dictionary whose keys are l-values and values are scattering radii as a
+        function of energy
+
+    Attributes
+    ----------
+    channel_radius : dict
+        Dictionary whose keys are l-values and values are channel radii as a
+        function of energy
+    energy_max : float
+        Maximum energy of the resolved resonance range in eV
+    energy_min : float
+        Minimum energy of the resolved resonance range in eV
+    scattering_radius : dict
+        Dictionary whose keys are l-values and values are scattering radii as a
+        function of energ
+    target_spin : float
+        Intrinsic spin, :math:`I`, of the target nuclide
+
+    """
+    def __init__(self, target_spin, energy_min, energy_max, channel, scattering):
+        self.target_spin = target_spin
+        self.energy_min = energy_min
+        self.energy_max = energy_max
+        self.channel_radius = channel
+        self.scattering_radius = scattering
+
+        self._prepared = False
+        self._parameter_matrix = {}
+
+    @classmethod
+    def from_endf(cls, ev, file_obj, items):
+        """Create resonance range from an ENDF evaluation.
+
+        This factory method is only used when LRU=0, indicating that only a
+        scattering radius appears in MF=2, MT=151. All subclasses of
+        ResonanceRange override this method with their own.
+
+        Parameters
+        ----------
+        ev : openmc.data.endf.Evaluation
+            ENDF evaluation
+        file_obj : file-like object
+            ENDF file positioned at the second record of a resonance range
+            subsection in MF=2, MT=151
+        items : list
+            Items from the CONT record at the start of the resonance range
+            subsection
+
+        Returns
+        -------
+        openmc.data.ResonanceRange
+            Resonance range data
+
+        """
+        energy_min, energy_max = items[0:2]
+
+        # For scattering radius-only, NRO must be zero
+        assert items[4] == 0
+
+        # Get energy-independent scattering radius
+        items = get_cont_record(file_obj)
+        target_spin = items[0]
+        ap = Polynomial((items[1],))
+
+        # Calculate channel radius from ENDF-102 equation D.14
+        a = Polynomial((0.123 * (1.008665*ev.target['mass'])**(1./3.) + 0.08,))
+
+        return cls(target_spin, energy_min, energy_max, {0: a}, {0: ap})
+
+    def reconstruct(self, energies):
+        """Evaluate cross section at specified energies.
+
+        Parameters
+        ----------
+        energies : float or Iterable of float
+            Energies at which the cross section should be evaluated
```

Comment:
I never said that it wouldn't be difficult, only that it wouldn't be "as bad" as the Fortran side :smile: All the complexities you've listed are among the reasons I haven't yet taken a stab at it.

For unit conversion, it will definitely take more than 20 LoC (not sure how much more, maybe ~100 LoC?). I'd need to convert units on:
- Reaction cross sections
- Reaction Q-values
- Reaction product yields
- Photon production yields / cross sections
- Prompt, total, and delayed nu (since they appear separately from other product yields)
- Angular distributions
- Energy distributions
- Fission energy release (although this might be net negative LoC?)
- Resonance parameters (and the reconstruction functions, since they're already written)

For distributions, it's often necessary to change both x- and y-values of a function since x is in units of energy and y is in units of inverse energy.

If I'm going to go through the effort of making units consistent, I'd actually rather just convert ACE data to eV rather than ENDF data to MeV, but that would have to be predicated on others' willingness to move to eV units. If you're open to that, I'll probably do it as a separate PR (which can be reviewed/merged prior to this one if you're worried about unit consistency).